File: C++/AddBinary.cc
// Given two binary strings, return their sum (also a binary string).

// For example,
// a = "11"
// b = '1'
// Return "100"

class Solution {
public:
    string addBinary(string a, string b) {
        string output;
        int carry = 0;
        auto ait = a.rbegin();
        auto bit = b.rbegin();
        while(ait != a.rend() || bit != b.rend() || carry) {
            int ai = ait == a.rend() ? 0 : *ait++ - '0';
            int bi = bit == b.rend() ? 0 : *bit++ - '0';
            int res = ai + bi + carry;
            output.push_back('0' + res % 2);
            carry = res / 2;
        }
        reverse(output.begin(), output.end());
        return output;
    }
};

File: Java/AddBinary.java
// Given two binary strings, return their sum (also a binary string).

// For example,
// a = "11"
// b = "1"
// Return "100".

public class AddBinary {
    public String addBinary(String a, String b) {
        int carry = 0;
        int i = a.length() - 1, j = b.length() - 1;
        StringBuilder sb = new StringBuilder();
        while (i >= 0 || j >= 0 || carry == 1) {
            int l = 0, r = 0;
            if (i >= 0) {
                l = a.charAt(i--) - '0';
            }
            if (j >= 0) {
                r = b.charAt(j--) - '0';
            }
            int res = l + r + carry;
            sb.append(res % 2);
            carry = res / 2;
        }
        return sb.reverse().toString();
    }
}
